cdap-app-templates/cdap-etl/cdap-data-streams-base/src/main/java/io/cdap/cdap/datastreams/SparkStreamingPipelineRunner.java

```diff
@@ -185,4 +214,124 @@ protected SparkCollection<Object> handleJoin(Map<String, SparkCollection<Object>
+            // it will create a new function for each RDD, which would limit each RDD but not the entire DStream.
+            batchRDD = new LimitingFunction<>(spec.getNumOfRecordsPreview()).call(batchRDD);
+          }
+          batchRDD = new CountingTransformFunction<>(stageSpec.getName(), sec.getMetrics(), "records.out", dataTracer)
+            .call(batchRDD);
+          JavaRDD<RecordInfo<Object>> wrapped = batchRDD.map(new WrapOutputTransformFunction<>(stageSpec.getName()));
+          StreamingContext streamingContext = new DefaultStreamingContext(stageSpec, sec, javaStreamingContext);
+          RDDCollection<RecordInfo<Object>> rddCollection =
+            new RDDCollection<RecordInfo<Object>>(sec, functionCacheFactory, javaStreamingContext.sparkContext(),
+                                                  new SQLContext(javaStreamingContext.sparkContext()),
+                                                  context, new SparkBatchSinkFactory(), wrapped);
+          EmittedRecords records = getEmittedRecords(pipelinePhase, stageSpec, rddCollection, groupedDag,
+                                                      branchers, shufflers, false, false);
+          Map<String, EmittedRecords> emittedRecords = new HashMap<>();
+          Collection<Runnable> sinkRunnables = new ArrayList<>();
+          emittedRecords.put(source, records);
+
+          //process the remaining stages
+          for (int i = 1; i < topologicalOrder.size(); i++) {
+            String stageName = topologicalOrder.get(i);
+            processStage(phaseSpec, sourcePluginType, sec, stagePartitions, pluginContext, collectors, pipelinePhase,
+                         functionCacheFactory, macroEvaluator, emittedRecords, groupedDag, groups, branchers, shufflers,
+                         sinkRunnables, stageName, time.milliseconds(), context);
+          }
+
+          //We should have all the sink runnables at this point, execute them
+          executeSinkRunnables(sec, sinkRunnables);
+
+          if (dStream instanceof StreamingEventHandler) {
+            ((StreamingEventHandler) dStream).onBatchCompleted(streamingContext);
+          } else if (dStream.dstream() instanceof StreamingEventHandler) {
+            ((StreamingEventHandler) (dStream.dstream())).onBatchCompleted(streamingContext);
+          }
+
+        }, Exception.class);
+
+      });
+
+    } else {
+      super.processDag(phaseSpec, sourcePluginType, sec, stagePartitions, pluginContext, collectors, pipelinePhase,
+                       functionCacheFactory, macroEvaluator, groupedDag, groups, branchers, shufflers);
+    }
+  }
+
+  @Override
+  protected Runnable getSparkSinkRunnable(StageSpec stageSpec, SparkCollection<Object> stageData,
+                                          SparkSink<Object> sparkSink, long time,
+                                          @Nullable DatasetContext datasetContext) throws Exception {
+    if (nativeStateStoreEnabled) {
```

Comment:
similar comment to above, normally better to put the short part in the check:

```
if (!...) {
  return super...
}
...
```

Reply:
Done